### PR TITLE
Add DirectWriteForwarder unit tests

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
Using [Visual Studio's Code Coverage tool](https://learn.microsoft.com/visualstudio/test/microsoft-code-coverage-console-tool?view=visualstudio), reported code coverage is **71.7 %**.

Created test suite for the DirectWriteForwarder project:

Test coverage includes:
- Font, FontFace, FontFamily, FontCollection classes
- All enums and structs (FontWeight, FontStretch, FontStyle, etc.)
- FontMetrics and GlyphMetrics calculations
- DWriteMatrix transforms (identity, scale, rotation)
- LocalizedStrings IDictionary implementation
- FontFile, FontSource, FontSourceFactory infrastructure
- FontFileLoader, FontFileStream, FontFileEnumerator
- ItemProps, Span, and ClassificationUtility for text analysis
- TrueTypeSubsetter.ComputeSubset with unsafe code
- Unsafe pointer APIs (TryGetFontTable, GetGlyphIndices, GetGlyphMetrics)
- TextAnalyzer integration tests (Itemize, GetGlyphs, GetGlyphPlacements)
- Edge cases and error handling (boundary conditions, invalid inputs)

Infrastructure changes:
- Added InternalsVisibleTo for DirectWriteForwarder.Tests in DirectWriteForwarder
- Added InternalsVisibleTo for DirectWriteForwarder.Tests in PresentationCore
- Created test project targeting net10.0-windows with xUnit and FluentAssertions
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/wpf/pull/11381)

### CODE REVIEW

**Comment by AaronRobinsonMSFT** [comment]
I'm currently done with this PR. The highlevel is the tests appear to be passing on all legs, look for `DirectWriteForwarder.Tests`, the reported code coverage is 71.1 %, and the only product change is adding the `InternalsVisibleTo` attribute to the `DirectWriteForwarder` and `PresentationCore` for the `DirectWriteForwarder.Tests` project.

/cc @dipeshmsft @JeremyKuhne @agocke @jkoritzinsky

File: src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/TrueTypeSubsetterTests.cs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+                IndexOutOfRangeException? caught = null;
+                try
+                {
+                    TrueTypeSubsetter.ComputeSubset(pFontData, fontData.Length, arialUri, 0, glyphArray);
+                }
+                catch (IndexOutOfRangeException ex)
+                {
+                    caught = ex;
+                }
+                Assert.NotNull(caught);
```
Comment (jkoritzinsky) [inline src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/TrueTypeSubsetterTests.cs]: Convert to `Assert.Throws`?

File: src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/TrueTypeSubsetterTests.cs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        {
+            fixed (byte* pFontData = fontData)
+            {
+                var result = TrueTypeSubsetter.ComputeSubset(pFontData, fontData.Length, arialUri, 0, glyphArray);
+
+                result.Should().NotBeNull();
+                result.Should().NotBeEmpty();
+            }
+        }
+    }
```
Comment (jkoritzinsky) [inline src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/TrueTypeSubsetterTests.cs:205]: This feels like it's not really validating the results of the test, just that there's data returned at all.

File: src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/DWriteTypeConverterTests.cs [hunk 3]
```diff
@@ -0,0 +1,841 @@
...
+    [InlineData(7)]  // Expanded
+    [InlineData(8)]  // ExtraExpanded
+    [InlineData(9)]  // UltraExpanded
+    public void Convert_FontStretch_ToNative_ShouldSucceed(int stretchValue)
+    {
+        var arialFamily = TestHelpers.GetArialFamilyOrSkip();
+
+        var stretch = (FontStretch)stretchValue;
+        var matchingFonts = arialFamily.GetMatchingFonts(FontWeight.Normal, stretch, FontStyle.Normal);
+        matchingFonts.Should().NotBeNull();
```
Comment (jkoritzinsky) [inline src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/DWriteTypeConverterTests.cs:96]: Could we have this test validate that the correct matching font for a given weight, stretch, and style is returned instead of "any font"? (multiple tests)

File: src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/DWriteTypeConverterTests.cs [hunk 4]
```diff
@@ -0,0 +1,841 @@
...
+    public void Convert_InformationalStringID_ToNative_ShouldSucceed(int stringIdValue)
+    {
+        var arialFamily = TestHelpers.GetArialFamilyOrSkip();
+        Assert.SkipUnless(arialFamily.Count > 0, "Arial font family has no fonts");
+
+        var font = arialFamily[0u];
+        var stringId = (InformationalStringID)stringIdValue;
+
+        // GetInformationalStrings calls DWriteTypeConverter.Convert(InformationalStringID)
+        _ = font.GetInformationalStrings(stringId, out _);
```
Comment (jkoritzinsky) [inline src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/DWriteTypeConverterTests.cs]: Have the expected informational strings for a given font hard-coded in the test?

**Comment by ThomasGoulet73** [comment]
@AaronRobinsonMSFT I'm a bit curious with the background for this PR, is it because you're planning on working on DirectWriteForwarder or is it just to increase code coverage (Maybe to increase testing of C++/CLI for the .Net interop team) ?

I ask this because I have open PRs (dotnet/wpf#9902 and dotnet/wpf#10523, though they've been open for a while) and more code locally to migrate DirectWriteForwarder to C# and I just want to make sure I'm not stepping on anybody's toes.

In any case, this PR would help a lot with migrating DirectWriteForwarder to C# so that's very nice.

File: src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/TrueTypeSubsetterTests.cs (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/TrueTypeSubsetterTests.cs]: Can't because of pinning.

File: src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/DWriteTypeConverterTests.cs (hunk 3 above)
Comment (AaronRobinsonMSFT) [reply src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/DWriteTypeConverterTests.cs:96]: So that is possible, but not really the point for this PR or library at present. This is about strict code coverage and that is satisfied sicne much of the APIs here are about wrapping native APIs. The correctness angle is important for sure, but not precisely the point given this is wrapping calls to `IDWriteFontFamily`.

I'll go through this and see what APIs are doing more logic compared to just forwarding.

File: src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/TrueTypeSubsetterTests.cs (hunk 2 above)
Comment (AaronRobinsonMSFT) [reply src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/TrueTypeSubsetterTests.cs:205]: Agree. This can be made more useful.

File: src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/DWriteTypeConverterTests.cs (hunk 3 above)
Comment (jkotas) [reply src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/DWriteTypeConverterTests.cs:96]: If you are not validating the values returned by the APIs being tested, it means that the wrapper can be mixing up or corrupting the values being forwarded and the tests won't notice...

**Comment by AaronRobinsonMSFT** [comment]
> more code locally to migrate DirectWriteForwarder to C#

@ThomasGoulet73 This is indeed the long term goal, but we need tests. That is the purpose/beginning of this PR process. We need measurable metrics in order to have confidence in any port to C# and the cost of doing it manually is non-zero. The current suite of WPF integration and unit tests is helpful but not sufficient for something as low level and sensitive as the `DirectWriteForwarder` assembly. Regardless if/when moving toward C# or not, the tests help us have confidence in changes so that is a win I'm assuming we can all agree on.

> In any case, this PR would help a lot with migrating DirectWriteForwarder to C# so that's very nice.

Yep.

File: src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/DWriteTypeConverterTests.cs (hunk 3 above)
Comment (AaronRobinsonMSFT) [reply src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/DWriteTypeConverterTests.cs:96]: I'm not disagreeing that the tests here are incomplete or not as robust as needed. I'm simply pointing out that the metric being optimized for in this pass in coverage. The boiler plate is now here and additional testing can now be added in a more targeted fashion. That is not the point of this particular pass, more needs to be done.

**Comment by ThomasGoulet73** [comment]
> This is indeed the long term goal, but we need tests. That is the purpose/beginning of this PR process. We need measurable metrics in order to have confidence in any port to C# and the cost of doing it manually is non-zero. The current suite of WPF integration and unit tests is helpful but not sufficient for something as low level and sensitive as the `DirectWriteForwarder` assembly. Regardless if/when moving toward C# or not, the tests help us have confidence in changes so that is a win I'm assuming we can all agree on.

That's good news, thanks

File: src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/TrueTypeSubsetterTests.cs (hunk 1 above)
Comment (jkoritzinsky) [reply src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/TrueTypeSubsetterTests.cs]: Could we wrap the whole fixed block in the Throws callback?

File: src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/DWriteTypeConverterTests.cs (hunk 3 above)
Comment (jkoritzinsky) [reply src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/DWriteTypeConverterTests.cs:96]: Coverage without validation feels to me like false coverage. The number looks good but it doesn't actually give us confidence for changes. I'd rather not have these tests that don't provide real validation in this first pass and add them in a second PR.

Comment (AaronRobinsonMSFT) [reply src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/DWriteTypeConverterTests.cs:96]: This isn't about a traditional coverage metric and "false coverage" isn't the argument here. The goal of this is to have something to build on top of. Understanding how to exercise the code is the point and the course coverage metric is the goal. If this was a code base that is having active development, I fully agree with this concern and false coverage is a concern. This conversation is missing the context of this repo and intent of the work.

I will take another stab and expanding these tests, but the only thing that is going to happen is increase the change and make for more to review and validate. We then play the probability game of seeing tests that are potentially insufficient. Regardless of anything that is checked in here, this is about boilerplate and these tests will need to be re-examined as code changes because everything here is new.

File: src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/TrueTypeSubsetterTests.cs (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/TrueTypeSubsetterTests.cs]: I suppose, what does that accomplish?

File: src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/TrueTypeSubsetterTests.cs (hunk 2 above)
Comment (AaronRobinsonMSFT) [reply src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/TrueTypeSubsetterTests.cs:205]: Done.

File: src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/TrueTypeSubsetterTests.cs (hunk 1 above)
Comment (jkoritzinsky) [reply src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/TrueTypeSubsetterTests.cs]: I think it's cleaner to read a test that's validating that an exception is thrown via `Assert.Throws` than having to mentally reconstruct that when it's done manually.

File: src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/DWriteTypeConverterTests.cs (hunk 3 above)
Comment (jkoritzinsky) [reply src/Microsoft.DotNet.Wpf/tests/UnitTests/DirectWriteForwarder.Tests/DWriteTypeConverterTests.cs:96]: If this is just intended to be "tests that exercise the paths that we can later expand with actual validation," sure. I wouldn't feel much more comfortable with making changes to DirectWriteForwarder with this PR than without given the quality of the tests in this PR, but if that's a non-goal for this PR, then it's fine.

**Review by jkoritzinsky — DISMISSED** [review]
Looks fine enough for the goal of this PR.

Would like to see the comment from the Copilot review applied before this is merged though.
